crates/bevy_math/src/rects/irect.rs

```diff
@@ -0,0 +1,426 @@
+use crate::IVec2;
+
+/// A rectangle defined by two opposite corners.
+///
+/// The rectangle is axis aligned, and defined by its minimum and maximum coordinates,
+/// stored in `IRect::min` and `IRect::max`, respectively. The minimum/maximum invariant
+/// must be upheld by the user when directly assigning the fields, otherwise some methods
+/// produce invalid results. It is generally recommended to use one of the constructor
+/// methods instead, which will ensure this invariant is met, unless you already have
+/// the minimum and maximum corners.
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub struct IRect {
+    /// The minimum corner point of the rect.
+    pub min: IVec2,
+    /// The maximum corner point of the rect.
+    pub max: IVec2,
+}
+
+impl IRect {
+    /// Create a new rectangle from two corner points.
+    ///
+    /// The two points do not need to be the minimum and/or maximum corners.
+    /// They only need to be two opposite corners.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_math::IRect;
+    /// let r = IRect::new(0, 4, 10, 6); // w=10 h=2
+    /// let r = IRect::new(2, 3, 5, -1); // w=3 h=4
+    /// ```
+    #[inline]
+    pub fn new(x0: i32, y0: i32, x1: i32, y1: i32) -> Self {
+        Self::from_corners(IVec2::new(x0, y0), IVec2::new(x1, y1))
+    }
+
+    /// Create a new rectangle from two corner points.
+    ///
+    /// The two points do not need to be the minimum and/or maximum corners.
+    /// They only need to be two opposite corners.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_math::{IRect, IVec2};
+    /// // Unit rect from [0,0] to [1,1]
+    /// let r = IRect::from_corners(IVec2::ZERO, IVec2::ONE); // w=1 h=1
+    /// // Same; the points do not need to be ordered
+    /// let r = IRect::from_corners(IVec2::ONE, IVec2::ZERO); // w=1 h=1
+    /// ```
+    #[inline]
+    pub fn from_corners(p0: IVec2, p1: IVec2) -> Self {
+        Self {
+            min: p0.min(p1),
+            max: p0.max(p1),
+        }
+    }
+
+    /// Create a new rectangle from its center and size.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if any of the components of the size is negative.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_math::{IRect, IVec2};
+    /// let r = IRect::from_center_size(IVec2::ZERO, IVec2::splat(2)); // w=2 h=2
+    /// assert_eq!(r.min, IVec2::splat(-1));
+    /// assert_eq!(r.max, IVec2::splat(1));
+    /// ```
+    #[inline]
+    pub fn from_center_size(origin: IVec2, size: IVec2) -> Self {
+        assert!(size.cmpge(IVec2::ZERO).all());
```

Comment:
Not sure if we only want this to be a `debug_assert` or not. I prefer keeping asserts out of potentially hot math code. Especially since we assert in `from_center_half_size` as well.